crates/core/src/host/host_controller.rs

```diff
@@ -170,27 +181,76 @@ impl HostController {
     /// system, the applications will expect to be called when they are scheduled to be called regardless
     /// of whether the OS has been restarted.
     pub async fn spawn_module_host(&self, mhc: ModuleHostContext) -> Result<ModuleHost, anyhow::Error> {
+        let res = self.start_module_host(mhc).await?;
+        self.commit_module_host(res).await
+    }
+
+    /// Set up the [`ModuleHost`] described by the provided [`ModuleHostContext]
+    /// and `start()` it.
+    ///
+    /// Do not, however, start the associated scheduler nor register the module
+    /// host with this host controller.
+    ///
+    /// This allows the module host to be discarded in case init / update
+    /// lifecycle routines fail.
+    ///
+    /// On success, pass the returned value to [`Self::commit_module_host`] to
+    /// complete the setup and register the host with this host controller.
+    async fn start_module_host(&self, mhc: ModuleHostContext) -> anyhow::Result<SpawnedModuleHost> {
         let key = mhc.dbic.database_instance_id;
 
+        let program_hash = hash_bytes(&mhc.program_bytes);
         let mcc = ModuleCreationContext {
-            dbic: mhc.dbic,
+            dbic: mhc.dbic.clone(),
             scheduler: mhc.scheduler,
-            program_hash: hash_bytes(&mhc.program_bytes),
+            program_hash,
             program_bytes: mhc.program_bytes,
             energy_monitor: self.energy_monitor.clone(),
         };
         let module_host = spawn_rayon(move || Self::make_module_host(mhc.host_type, mcc)).await?;
+        module_host.start();
+
+        Ok(SpawnedModuleHost {
+            key,
+            module_host,
+            scheduler_starter: mhc.scheduler_starter,
+            supplied_program: program_hash,
+            dbic: mhc.dbic,
+        })
+    }
 
-        let old_module = self.modules.lock().insert(key, module_host.clone());
+    /// Complete the setup of the [`ModuleHost`] started using [`Self::start_module_host`].
+    ///
+    /// Exits the previous version of the module, if any, starts the new module's
+    /// scheduler, and registers the module with this host controller.
+    async fn commit_module_host(&self, spawned: SpawnedModuleHost) -> anyhow::Result<ModuleHost> {
+        {
+            // At this point, the supplied program must be the program stored
```

Comment:
Yeah, in the future, I think `spawn_module_host` should read the program bytes from the database/commit log as you suggest. The way we could perhaps do it would be to send a message to the host_controller to stop all modules, then insert a new row in the datastore directly and then send a message to the host_controller to reload the modules.

Perhaps there's an even better way to do it, that's just off the top of my head.